### PR TITLE
ENH: Improve ref-tracking for group keys

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jorisvandenbossche 

Follow-up on https://github.com/pandas-dev/pandas/pull/49450

The equals check causes failures in Dask. I think we are better off moving to reference checking anyway. If you are ok with the general logic here, I'd refactor this check down to the Manager level.

### CODE REVIEW

File: pandas/core/groupby/grouper.py [hunk 1]
```diff
@@ -947,9 +948,11 @@ def is_in_obj(gpr) -> bool:
             # For the CoW case, we need an equality check as the identity check
             # no longer works (each Series from column access is a new object)
             try:
-                return gpr.equals(obj[gpr.name])
+                obj_gpr_column = obj[gpr.name]
             except (AttributeError, KeyError, IndexError, InvalidIndexError):
                 return False
+            ref = weakref.ref(obj_gpr_column._mgr.blocks[0])
+            return ref in gpr._mgr.blocks[0].refs.referenced_blocks
```
Comment (rhshadrach) [inline pandas/core/groupby/grouper.py:955]: I haven't checked yet, but I think the phenomenon in #50805 could mean `obj[grp.name]` might return a DataFrame.

Comment (rhshadrach) [reply pandas/core/groupby/grouper.py:955]: Ah - no - I think this is okay. This should always return a Series (we check provided groupers are 1d).

**Comment by rhshadrach** [comment]
This would make `df.groupy(df['a'])` behave differently from `df.groupby(df['a'].copy())`, is that right? In general I'd like to move away from "in-axis" vs "out-axis" groupers behaving differently (ref: #49519), so I agree this is the right direction, but it might cause some odd edge cases to behave differently until we resolve those issues.

Still - I'm +0 here; checking equality gives some odd behaviors as well.

**Comment by phofl** [comment]
Yes those 2 behave different, but this pr establishes the same behavior as in the non-cow mode for this case. So that's actually a plus imo

File: pandas/core/groupby/grouper.py [hunk 2]
```diff
@@ -947,9 +948,13 @@ def is_in_obj(gpr) -> bool:
             # For the CoW case, we need an equality check as the identity check
             # no longer works (each Series from column access is a new object)
             try:
-                return gpr.equals(obj[gpr.name])
+                obj_gpr_column = obj[gpr.name]
             except (AttributeError, KeyError, IndexError, InvalidIndexError):
                 return False
+            if isinstance(gpr, Series) and isinstance(obj_gpr_column, Series):
+                ref = weakref.ref(obj_gpr_column._mgr.blocks[0])
+                return ref in gpr._mgr.blocks[0].refs.referenced_blocks
```
Comment (jorisvandenbossche) [inline pandas/core/groupby/grouper.py:956]: So if you create multiple times a weakref to the same object, this is always the same (i.e. identical) weakref object?

**Comment by jorisvandenbossche** [comment]
Yes, I think this is a good solution for now (essentially the same as what was suggested in https://github.com/pandas-dev/pandas/pull/50730). And it indeed aligns with the behaviour we have for non-CoW.

Longer term, I agree with @rhshadrach that the current semantics are not really great. We should probably at some point move to not doing such check at all, at which point this discussion about identity vs equality also just doesn't matter anymore?

**Comment by phofl** [comment]
Yep this sounds sensible, but I think this needs a deprecation cycle?

I’ll move the check to the BlockManager then

**Comment by phofl** [comment]
Implemented it on the manager level

**Review by jorisvandenbossche — APPROVED** [review]
Just two small comments to clean-up

File: pandas/core/groupby/grouper.py [hunk 3]
```diff
@@ -947,9 +947,19 @@ def is_in_obj(gpr) -> bool:
             # For the CoW case, we need an equality check as the identity check
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/core/groupby/grouper.py]: Done

File: pandas/core/groupby/grouper.py [hunk 4]
```diff
@@ -947,9 +947,19 @@ def is_in_obj(gpr) -> bool:
...
             # no longer works (each Series from column access is a new object)
             try:
-                return gpr.equals(obj[gpr.name])
+                obj_gpr_column = obj[gpr.name]
             except (AttributeError, KeyError, IndexError, InvalidIndexError):
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/core/groupby/grouper.py]: Correct, thx

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**